### PR TITLE
details pane overview refactoring

### PR DESCRIPTION
**Pull request by 0xF013**
[`* `#17195 ` boolean`](http://community.openproject.org/work_packages/17195)
[`* `#17194 ` textarea`](http://community.openproject.org/work_packages/17194)
[`* `#17217 ` list`](http://community.openproject.org/work_packages/17217)
[`* `#17216 ` text, integer, float`](http://community.openproject.org/work_packages/17216)
[`* `#17218 ` version`](http://community.openproject.org/work_packages/17218)
[`* `#16643 ` user`](http://community.openproject.org/work_packages/16643)
[`* `#16944 ` budget`](http://community.openproject.org/work_packages/16944)
- [x] ?maybe switch version from another request to embedded check
- [x] spent units not counted as empty
- [x] revert height change on edit mode
- [x] make users links on author (also check for activity)
- [x] fix description format check
- [x] subject width changes in edit mode
- [x] non-textile description top covering Description header
- [x] remainingHours
- [x] estimatedTime
- [ ]  separate pr
  - [ ] date
  - [ ] date range
  - [ ] type
  - [ ] activity textarea onleave
- [x] enforce CF sorting
- [x] check for defining project when displaying version link
  - [x] extract version field in a separate directive (?)
-  [x] check unavailable version label
- [x] check obsolete translations
- [x] enable and fix backlogs and budget fields
  - [x] fix aggregated budget field
  - [x] create a PR in the costs plugin rep
- [x] deliver errors to all editables
- [x] deliver 'close' to all editables
- [x] fix version view for verion CF
- [x] boolean CF view
  - [x] check save with @NobodysNightmare 
  - [x] focus lost
- [x] input type number for integer and float CF
- [x] fix show all
- [x] refactor schema request in routing resolves and workpackage refresh
- [x] divide inplace editor dispatcher into several components
- [ ] integration tests
  - [ ] integration tests for time rounding to 2 digits after point
- [x] remove commented code
- [x] fix window refresh confirmation message from textarea
- [x] hide buttons during preview
- [x] description placeholder
- [x] fix clicking on a link on preview / display
- [x] remove unused components
- [x] fix CF list with @NobodysNightmare 
- [x] hours should be a link
- [x]  check all plugins
- [x] allow setting to null fields with no values
- [x] update activity
- [x] version link not clickable
- [x] show stuff for logged out
- [x] fix focusable element when tabbing
- [x] fix styling of dashboard covering next inplace editor

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://0xf013-feature-inplace-editable-custom-fields.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://0xf013-feature-inplace-editable-custom-fields.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by 0xF013** [comment]
making a pull request now for visibility of the progress

**Comment by ulferts** [comment]
@myabc, @floriank can you please have a look on it ASAP as I think there might be other PRs coming along that will depend on it.

**Comment by 0xF013** [comment]
@ulferts it's nowhere close to being finished

**Comment by ulferts** [comment]
I am aware of that, just want to make sure that it gets premium PR handling.

**Comment by myabc** [comment]
@ulferts _premium PR handling_? sounds like something we could charge extra for.

**Comment by NobodysNightmare** [comment]
Just for the record:

I tested the saving of boolean custom fields and it works for me.

@0xF013 It is still possible, that you discovered a broken edge case, so please tell me how your broken custom field looks like (i.e. what are its settings? Is it required? Has it a default value?)

I'll have another look into that then, but for now I assume that your frontend code is doing it right.

File: frontend/app/work_packages/controllers/work-packages-controller.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-//-- copyright
+ //-- copyright
```
Comment (myabc) [inline frontend/app/work_packages/controllers/work-packages-controller.js]: @0xF013 mind the extra whitespace

**Comment by NobodysNightmare** [comment]
Not sure if this is expected to work, so tell me and I will make a separate ticket for that:

If you set your OP instance to no formatting (i.e. **not** textile), I'd expect the description to not have formatting controls, but they are present.

As a client you can recognize the current format setting by looking at the `format` property of the description (or any other `Formattable` you are interested in): If ti is `plain`, there is no formatting available, otherwise there is (and we currently only support `textile` as other option).

**Comment by 0xF013** [comment]
@NobodysNightmare that is a bug, thanks

**Comment by NobodysNightmare** [comment]
On the styling side:
I remember the inplace edit controls (✓, ✓ ✉ and ❌) to be placed absolutely, thereby _not increasing the height of their parent box_.

Currently opening the edit controls leads to the parent box growing in height and thereby destroying the layout.

Without controls:

Baseline exactly aligned

With controls:

Baseline not nearly aligned

**Comment by 0xF013** [comment]
@NobodysNightmare I added the increase in edit mode because for two inputs that are edited simultaneously one after another, the control buttons overlap with the next row and stuff is not clickable.

**Comment by NobodysNightmare** [comment]
You mean like that:
layout-dev

After a fix provided by @ulferts a while ago, the bottom element is still clickable where it is not visually obscured (e.g. I can still change the Version). So **I would** say the behaviour is known and intended, but that is only my impression.

Did your changes introduce a new case where a property is completely unusable when the element above it has its controls expanded?

**Comment by 0xF013** [comment]
@NobodysNightmare I will revert it and let's see

**Comment by NobodysNightmare** [comment]
Okay, I'll retest the frontend once you've done that. Code-wise I did not have a look, but the functionality looks fine :)

**Comment by NobodysNightmare** [comment]
@0xF013 I see that there can be problematic cases like this one:

overlay

Still I think it would be a change of known behaviour to change these styles here. I'd say let's discuss that separately from this PR, if you are fine with that...

**Comment by NobodysNightmare** [comment]
A styling issue that is new in your PR.

RTA description:

description-title-visible

plain description:

description-title-hidden

In the latter case the title (**DESCRIPTION**) is hidden (probably by the textarea).
I am not sure if this is caused by the description specific CSS hack that brings the rich-edit controls into a single line with the description...

**Comment by 0xF013** [comment]
@NobodysNightmare description and subject fixes pushed

**Comment by NobodysNightmare** [comment]
### Reproduction
1. enter comment into details pane
2. submit comment
3. refresh page (e.g. click into address bar and press enter)

### Expected

page simply refreshes

### Actual

warning, that I have unsaved changes:

unsaved-changes

**Comment by NobodysNightmare** [comment]
I found a backend-side issue regarding custom fields, that is quite dubious.

If you encounter any of the following it is an API issue not related to the changes by Mihail:
- a bool CF set to false is deleted when you change an unrelated property
- a bool CF is deleted when you try to set it to false
- when you add a comment, a deleted bool CF is added as false

Weird Weird...

**Comment by 0xF013** [comment]
I sense some falsey circus

**Comment by 0xF013** [comment]
@NobodysNightmare regarding onunload handler, it worked as that before my PR. I would fix that in a separate workpackage

File: app/assets/stylesheets/content/_in_place_editing.sass [hunk 2]
```diff
@@ -5,7 +5,7 @@ $inplace-edit--color--very-dark: #cacaca
 .inplace-edit
   @extend .form--field.-full-width
 
-  &.-preview
+  .-preview
```
Comment (myabc) [inline app/assets/stylesheets/content/_in_place_editing.sass:8]: @0xF013 is this correct?

Comment (0xF013) [inline app/assets/stylesheets/content/_in_place_editing.sass:8]: @myabc I moved the preview state on an element lower.

**Comment by NobodysNightmare** [comment]
Upon opening the details pane it seems like the focus slides along all editable fields briefly before finally coming to a halt at the subject.

This behaviour is new and hard to describe... Do you need a screencast or can you reproduce it yourself?

embedded GIF
[link to GIF if embedding is broken](https://gfycat.com/DeadlyRevolvingArieltoucan)

**Comment by NobodysNightmare** [comment]
Not important, but if it is easy peasy:

Using the arrows of the input for the estimated time I can spin to negative values. It would be great if it stopped at 0.

**Note**: Really just a _nice-to-have_. Do not put any effort into this if it is more than e.g. adding `minvalue="0"`

**Comment by myabc** [comment]
> Not important, but if it is easy peasy:

@NobodysNightmare It's so easy-peasy, you could probably do it :)
`frontend/app/templates/work_packages/inplace_editor/custom/editable/duration.html`

The attribute is `min="0"` not `minvalue="0"`. See http://www.w3.org/TR/html-markup/input.number.html

File: app/assets/stylesheets/content/_in_place_editing.sass (hunk 2 above)
Comment (myabc) [inline app/assets/stylesheets/content/_in_place_editing.sass:8]: @0xF013 modifier-type classes should never be used standalone – but we can adjust this in another PR / @floriank

File: frontend/app/work_packages/controllers/details-tab-overview-controller.js [hunk 3]
```diff
@@ -26,325 +26,64 @@
...
+           WorkPackageFieldService
+           ) {
 
-  $scope.inplaceProperties = OverviewTabInplaceEditorConfig.getInplaceProperties();
+  var vm = this;
 
-  $scope.userPath = PathHelper.staticUserPath;
-  AuthorisationService.initModelAuth('work_package' + $scope.workPackage.id,
-                                     $scope.workPackage.links);
+  vm.groupedFields = [];
```
Comment (myabc) [inline frontend/app/work_packages/controllers/details-tab-overview-controller.js:37]: :+1:

**Pull request merged**